### PR TITLE
refactor(iroh-base)!: reduce dependencies

### PR DESCRIPTION
**Pull request by dignifiedquire**
## Description

- ensure `zeroize` is used for `SecretKey`
- remove unused dependencies and features from `iroh-base`

## Breaking Changes

- `iroh_base::SecretKey::generate` now takes an rng
- removed `iroh_base::SecretKey::generate_with_rng`, use `generate` directly
- removed `iroh_base::SecretKey::to_openssh`
- removed `iroh_base::SecretKey::from_openssh`
- `anyhow::Error` is replaced with explicit errors for `RelayUrl::from_str`
- `anyhow::Error` is replaced with explicit errors for `SharedSecret::open`

## Notes & open questions

As a follow up, all usage of `SecretKey::generate` in tests should be changed to use a seeded rng, like chacha8, instead of `thread_rng`

## Change checklist

- [ ] Self-review.
- [ ] Documentation updates following the [style guide](https://rust-lang.github.io/rfcs/1574-more-api-documentation-conventions.html#appendix-a-full-conventions-text), if relevant.
- [ ] Tests if relevant.
- [ ] All breaking changes documented.

### CODE REVIEW

**Comment by matheus23** [comment]
Hm. I would argue the `SecretKey::generate` function change may not be justified. This increases the burden on the library user. We need to *at least* have a doctest that shows them how to use the right rng with `Secret::generate` (I'd say for most cases that's likely `rand::thread_rng()`).
I suspect that removing the `rand` dependency from `iroh-base` only makes `iroh-base` lighter, not `iroh` itself, does it?

File: iroh-base/src/key.rs
Comment (rklaehn) [inline iroh-base/src/key.rs:283]: This wasn't stable back when we wrote this initially, right?

File: iroh-base/src/key.rs
Comment (rklaehn) [inline iroh-base/src/key.rs:354]: These are useful. Can we have this code somewhere, maybe in examples?

**Review by matheus23 — APPROVED** [review]
Alright. I'll defer to your experience in that case then, but I *do* demand a `no_run` doctest with `rand::thread_rng()` example usage in the `SecretKey::generate` docs before you merge :P

File: iroh-base/src/key.rs
Comment (rklaehn) [inline iroh-base/src/key.rs:488]: "As a follow up, all usage of SecretKey::generate in tests should be changed to use a seeded rng, like chacha8, instead of thread_rng"

So what's wrong with using thread_rng in tests? Nondeterminism, or something else. I agree that nondeterminism in tests is not great, but we have not been religious about this in the past.

**Comment by dignifiedquire** [comment]
add an actually running doctest

**Review by rklaehn — APPROVED** [review]
Not having generate without providing a RNG is great.

File: iroh-base/src/key.rs
Comment (matheus23) [reply iroh-base/src/key.rs:488]: I think it's worth taking steps towards determinism.

File: iroh-base/src/key.rs
Comment (flub) [inline iroh-base/src/key.rs:344]: The rand crate is on version 0.8.  Is it really a good idea to require it on the public API?

File: iroh-base/src/relay_url.rs
Comment (flub) [inline iroh-base/src/relay_url.rs:41]: I realise this is bikeshedding, but it's somewhat surprising that this is "Pars**ing**", rather than `RelayUrlParseError`

File: iroh-base/src/key.rs
Comment (flub) [inline iroh-base/src/key.rs:342]: From your later changes seems like you can directly call `SecretKey::generate(rand::rngs::OsRng)` which might be a little simpler as recipe?

File: iroh-base/src/key.rs
Comment (dignifiedquire) [reply iroh-base/src/key.rs:283]: I think so

File: iroh-base/src/key.rs
Comment (dignifiedquire) [reply iroh-base/src/key.rs:354]: I would put this into the `load_key` function, which uses this, when we update iroh-blobs

File: iroh-base/src/key.rs
Comment (dignifiedquire) [reply iroh-base/src/key.rs:488]: it's only about nondeterminism why I want to do this, we already have a lot of randomness from the network, so reducing it from factors like this is usually very helpful

File: iroh-base/src/key.rs
Comment (dignifiedquire) [reply iroh-base/src/key.rs:344]: all of the rust crypto world uses `rand` like this in their public interfaces.. I don't think there is much point in trying to do sth differently

File: iroh-base/src/relay_url.rs
Comment (dignifiedquire) [reply iroh-base/src/relay_url.rs:41]: I don't care at all 🤷

File: iroh-base/src/key.rs
Comment (flub) [reply iroh-base/src/key.rs:488]: I kind off like the `SecretKey::from_bytes(&[1u8; 32])` approach for deterministic keys in tests.  But I agree that if you use only one key in a test I'm not sure why bother with it?

File: iroh-base/src/key.rs
Comment (dignifiedquire) [reply iroh-base/src/key.rs:342]: you can, but that's not what most code will end up doing, which is why I made it two explicit steps

File: iroh-base/src/relay_url.rs
Comment (dignifiedquire) [reply iroh-base/src/relay_url.rs]: changed

**Pull request merged**
